pkg/util/loop/loop.go

```diff
@@ -190,9 +202,16 @@ func (loop *Device) AttachFromFile(image *os.File, mode int, number *int) error
 			syscall.Syscall(syscall.SYS_IOCTL, uintptr(loopFd), CmdClrFd, 0)
 			return fmt.Errorf("failed to set loop flags on loop device: %s", syscall.Errno(err))
 		}
+
+		deviceFound = true
+
 		break
 	}
 
+	if eagainFound && !deviceFound {
+		return syscall.EAGAIN
+	}
+
 	loop.fd = new(int)
 	*loop.fd = loopFd
 	return nil
```

Comment:
Why don't we just set `loop.fd` and return nil within the loop (instead of the `break` on L208)?

If we do that, there is no need for the `deviceFound` flag, as the only possible reason that control would pass to the block after the loop would be when the device was not found. 

I also think there is no need for the `eagainFound` flag, as already it would be returning a specific error when a more permanent error is returned, so the only reason to have reached the block after the loop would be if we got EAGAIN/EBUSY.

At that point, our only interest is this block is to  convey to the caller whether or not it is worth retrying the call to this function again. I think returning `syscall.EAGAIN` is a somewhat hacky way to convey this, and it is a little bit dishonest since it could actually arise from EAGAIN or EBUSY. One option would be simply to return nil here (without setting `loop.fd` and the caller can check if `loop.fd == nil` to see if this function should be retried. 

Alternatively, it could make sense to add a custom type that implements the error interface and return that instead, and we can check the type of the error in the caller to test whether it is this error or one of the other (more permanent) ones. 

A custom error could be defined something like this:
```
type TransientAttachError struct{
    message string
}

func (tae *TransientAttachError) Error() string {
    return message
}
```

And then in this block after the device loop, simply return it:
```
return TransientAttachError{"failed to successfully allocate a loop device (please retry)"}
```

And then in the loop in `AttachFromFile`:
```
if err != nil {
    _, transient := err.(*TransientAttachError)
    if !transient {
        return err   
    }
    time.Sleep(250 * time.Millisecond)
}
```

Reply:
@jrandall I implemented most of your suggestions, but didn't manage to get rid of eagainFound after realizing that I need to check its value in the if device == loop.MaxLoopDevices { block, as before it would have been impossible to get beyond the for-loop when no suitable device was found but EAGAIN or EBUSY occurred at least once.